### PR TITLE
Symlink tests

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: utils/src/test/kotlin/ExtensionsTest.kt [hunk 1]
```diff
@@ -45,6 +45,40 @@ class ExtensionsTest : WordSpec({
...
+            dir.resolve("directory").safeMkdirs()
+            ProcessCapture(dir, "cmd", "/c", "mklink", "/d", "symlink-to-directory", "directory")
+            dir.resolve("symlink-to-directory").isSymbolicLink() shouldBe true
+            dir.safeDeleteRecursively()
+        }
+
+        "correctly detect hardlinks on Windows".config(enabled = Os.isWindows) {
+            val dir = createTempDir()
+            dir.resolve("file").createNewFile()
+            ProcessCapture(dir, "cmd", "/c", "mklink", "/f", "hardlink", "file")
```
Comment (fviernau) [inline utils/src/test/kotlin/ExtensionsTest.kt]: Is a `hardlink` actually a `symbolic link` on windows?

File: utils/src/test/kotlin/ExtensionsTest.kt [hunk 2]
```diff
@@ -45,6 +45,40 @@ class ExtensionsTest : WordSpec({
         }
     }
 
+    "File.isSymbolicLink" should {
```
Comment (fviernau) [inline utils/src/test/kotlin/ExtensionsTest.kt]: Consider moving these tests into a separate file, e.g. `FileExtensionsTest`.
As result you'd have tests with `similar setup` in a `separate file` and thus the test `setup` and `tearDown` can be moved into the dedicated methods which leads to less code redundancy an better readability.

File: utils/src/test/kotlin/ExtensionsTest.kt (hunk 1 above)
Comment (sschuberth) [reply utils/src/test/kotlin/ExtensionsTest.kt]: Ah, good point, no, it's a hardlink as also known from Linux. Let's see how the tests behave on AppVeyor, it should indeed fail. I currently cannot test this locally.

Comment (fviernau) [reply utils/src/test/kotlin/ExtensionsTest.kt]: My understanding of a `hardlink` under Linux is that roughly speaking you cannot distinguish between different hardlinks to the same file. That is a hardlink is just a reference from a file system (directory) node and a file exists as long as there is at least one `hardlink` pointing to it. 

I guess terminoligy of `hardlink` under Windows is analog - thus we'd just need to flip the `asserts expected value` ?!

Comment (sschuberth) [reply utils/src/test/kotlin/ExtensionsTest.kt]: While we could do that, I think it's clearer to just drop that test case.

Comment (sschuberth) [reply utils/src/test/kotlin/ExtensionsTest.kt]: Changed my mind, will even add more tests :-)

File: utils/src/test/kotlin/ExtensionsTest.kt [hunk 3]
```diff
@@ -45,6 +45,42 @@ class ExtensionsTest : WordSpec({
...
+            file.isSymbolicLink() shouldBe false
+        }
+
+        "not detect directories as symbolic links on Windows".config(enabled = Os.isWindows) {
+            directory.isSymbolicLink() shouldBe false
+        }
+
+        "not detect hard links as symbolic links on Windows".config(enabled = Os.isWindows) {
+            ProcessCapture(tempDir, "cmd", "/c", "mklink", "/f", "hardlink", "file")
+            tempDir.resolve("hardlink").isSymbolicLink() shouldBe false
```
Comment (fviernau) [inline utils/src/test/kotlin/ExtensionsTest.kt]: Consider putting an empty line above this one.

File: utils/src/test/kotlin/ExtensionsTest.kt [hunk 4]
```diff
@@ -45,6 +45,42 @@ class ExtensionsTest : WordSpec({
         }
     }
 
+    "File.isSymbolicLink" should {
+        val tempDir = createTempDir()
+        val file = tempDir.resolve("file").apply { createNewFile() }
+        val directory = tempDir.resolve("directory").apply { safeMkdirs() }
+
+        "not detect files as symbolic links on Windows".config(enabled = Os.isWindows) {
```
Comment (fviernau) [inline utils/src/test/kotlin/ExtensionsTest.kt]: Consider starting test names with: "return false for..."  (symbolic links on windows)

Consider also turning "symbolic links" into "symbolic file links" and the below one into "symbolic directory links"

**Comment by sschuberth** [comment]
Sorry, I need more time testing this and will rebase a few more times.

**Comment by sschuberth** [comment]
@fviernau @mnonnenmacher this is now ready for review again.

File: utils/src/test/kotlin/ExtensionsTest.kt (hunk 2 above)
Comment (sschuberth) [reply utils/src/test/kotlin/ExtensionsTest.kt]: Solved differently by moving setup and teardown code to the outer scope.

File: utils/src/test/kotlin/ExtensionsTest.kt (hunk 4 above)
Comment (sschuberth) [reply utils/src/test/kotlin/ExtensionsTest.kt]: I've done the former but not the latter.

File: utils/src/main/kotlin/Extensions.kt [hunk 5]
```diff
@@ -129,50 +129,26 @@ fun File.safeCopyRecursively(target: File, overwrite: Boolean = false) {
...
-                if (!force) throw e
-
-                if (file.toFile().setWritable(true)) {
-                    // Try again.
-                    Files.delete(file)
-                }
-            }
-
-            return FileVisitResult.CONTINUE
+    if (!delete() && force) {
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/Extensions.kt]: Done.

**Pull request merged**